### PR TITLE
QoL: Increase playercount to 8

### PR DESCRIPTION
**Pull request by kphoenix137**
image

### CODE REVIEW

File: Source/lighting.h [hunk 1]
```diff
@@ -21,7 +21,7 @@
 namespace devilution {
 
 #define MAXLIGHTS 32
-#define MAXVISION 4
+#define MAXVISION 8
```
Comment (glebm) [inline Source/lighting.h:24]: Let's set it to MAX_PLRS?

**Comment by AJenbo** [comment]
did you check what happens when players chat?

**Comment by ikonomov** [comment]
Dakk's meme comes to mind.  Just because we could doesn't mean it should.  I think this is a mod, not QoL.

**Comment by AJenbo** [comment]
I'm not sure lua would be able to do this since it's an engine limit.

**Comment by kphoenix137** [comment]
> did you check what happens when players chat?

Yes. nothing breaks. Just unable to mute the extra players. I think adding commands /mute and /unmute could be useful for this.

**Comment by kphoenix137** [comment]
> 8 seems like a random number.

Which number would not be a random number?

**Comment by kphoenix137** [comment]
Before undrafting, I'll also need to test out some other stuff, like town portal placement and golems. If anyone has ideas of anything that might break that I'm forgetting about, please share.

**Comment by AJenbo** [comment]
Basically the timedemo tests needs to pass for this to be merged, else I would have just merged https://github.com/diasurgical/devilutionX/pull/551 ;)

**Comment by kphoenix137** [comment]
Why does the timedemo fail? Isn't MAX_PLRS only used in multiplayer?

**Comment by kphoenix137** [comment]
> Basically the timedemo tests needs to pass for this to be merged, else I would have just merged #551 ;)

So we just need a new timedemo? Is anyone interested in doing this? Last time I tried I couldn't get it working.

**Comment by StephenCWills** [comment]
No, `MAX_PLRS` is used in Single Player and should therefore not be changed there or you'll break savegames. You need to make timedemo work without rerecording for this one.

**Comment by ikonomov** [comment]
> > 8 seems like a random number.
> 
> Which number would not be a random number?

https://github.com/diasurgical/devilutionX/pull/551 Something like 16.  It's as high as anybody would want to go.

**Comment by DakkJaniels** [comment]
16 seems random to me, e^pi seems much less random.

**Comment by AJenbo** [comment]
Looks like the original WoW had a limit of about 4500 players pr world instance so we should probably have at least that

**Comment by kphoenix137** [comment]
I mean ikonomov makes a point; no reason to use a random number. Let's just make it uncapped

**Comment by yuripourre** [comment]
This might be something stupid or maybe was fixed already but... IIRC, there is a limit of monsters, right? And because golems use the same array, if MAX_PLRS is too high it will allocate slots for golems and less monsters will be initialized, right?

**Comment by StephenCWills** [comment]
> This might be something stupid or maybe was fixed already but... IIRC, there is a limit of monsters, right? And because golems use the same array, if MAX_PLRS is too high it will allocate slots for golems and less monsters will be initialized, right?

You are absolutely correct. After this PR was submitted, I did some math to determine whether bumping the player count to 8 could have an effect on the number of monsters in a dungeon, and the answer is yes. However, even so, it didn't seem to me that occasionally having 4 fewer monsters would really be significant enough to bring it up. If we bumped the player count to 16, though, 12 fewer monsters is starting to sound significant.

> Edit: Btw, if golems were moved to a separate array or list, could mods create summons?

Mods can create summons regardless. But putting Golems in a separate collection might make the process of introducing new ones a bit easier, since much of the code for it would probably be copy/paste.

**Comment by ikonomov** [comment]
The fact that changing this limit reduces the number of monsters per level is significant.  Even if that number is just 4, in 16 levels that would be 64 less monsters.  If this is to be merged I think this issue needs to be resolved so that there is no change of monsters per level.  Also the mute functionality and the list of players in the control panel should be fully functional for 8 players, maybe it could be made scrollable.

Untitled-1

**Comment by StephenCWills** [comment]
> The fact that changing this limit reduces the number of monsters per level is significant. Even if that number is just 4, in 16 levels that would be 64 less monsters. If this is to be merged I think this issue needs to be resolved so that there is no change of monsters per level.

I could write a book on why this is a pretty flawed argument. But since you're advocating for no change at all on the basis that 4 fewer monsters is significant, I thought it might be more productive to simply ask how you'd feel about removing the upper limit of 186 monsters per dungeon level.

**Comment by ikonomov** [comment]
OK, let me rephrase.  It might not be significant, but it is not insignificant when considering a full level 1-16 run.  I'm not advocating for no change at all, I'm just suggesting that perhaps reducing the number of monsters is an undesirable effect.

**Comment by StephenCWills** [comment]
> OK, let me rephrase. It might not be significant, but it is not insignificant when considering a full level 1-16 run. I'm not advocating for no change at all, I'm just suggesting that perhaps reducing the number of monsters is an undesirable effect.

So it sounds like you are not outright rejecting the idea of removing that upper limit. But let me be clear.

186 monsters is the _upper limit_ on the number of monsters placed on each dlvl. Adding 4 more Golems reduces the _upper limit_ to 182, which _has the potential_ of reducing the total number of monsters by 4 in each dlvl. The actual number of monsters placed on each dlvl depends on the number of solid tiles in the dungeon's layout. I have no idea what a typical amount would be, and it probably varies a bit for each dungeon type, but the theoretical maximum (in an unrealistic scenario with a dungeon that has no solid tiles) before applying the upper limit is 213 in SP and 319 in MP.

**Comment by kphoenix137** [comment]
Since a change of this magnitude isn't in response to a bug, and is in arbitrary changes, I think its a good candidate for a Lua mod that can be included with the official installation

**Comment by AJenbo** [comment]
You can't really Lua you way out of engine limits like 4 player or 200 monsters.

**Comment by DakkJaniels** [comment]
Yeah you'd basically have to make all the changes to support changing the number (maybe two separate vectors for monsters and golems, and whatever other supporting changes would be needed), and then just having lua let you modify the number of players that can join the game or something - I don't think this is a trivial change.

**Comment by kphoenix137** [comment]
I think I have a solution for the mentioned issue, and also for fixing the problem with the single player file size being too large for reverse compatibility. Given that this change is specifically for multiplayer only, we can both allocate a monster array of size 204 for multiplayer only, while leaving it at 200 for single player, and set the max players to 8 in multiplayer only and leave it at 4 for single player. We can use the same arrays for both but only save up the vanilla limit for single player save files and ignore the rest.

**Comment by AJenbo** [comment]
Existing saves will have monsters in index 5-8 so you also need to change how golems are allocated.
Doing this proper will likely require some more cleanup and rework on various parts.

**Comment by kphoenix137** [comment]
> Existing saves will have monsters in index 5-8 so you also need to change how golems are allocated. Doing this proper will likely require some more cleanup and rework on various parts.

Good point. That will be tricky..

**Comment by kphoenix137** [comment]
> Existing saves will have monsters in index 5-8 so you also need to change how golems are allocated. Doing this proper will likely require some more cleanup and rework on various parts.

Specific to multiplayer?

**Comment by kphoenix137** [comment]
> Existing saves will have monsters in index 5-8 so you also need to change how golems are allocated. Doing this proper will likely require some more cleanup and rework on various parts.

The idea is to only allocate 4 slots for golems just like it currently works.  Only multiplayer would allocate 8 slots for golems.

**Comment by StephenCWills** [comment]
If we want to go beyond 8 players, though, we should probably take some of these suggestions seriously. Like turning the arrays into vectors and moving Golems to their own collection.

**Comment by AJenbo** [comment]
And a scrollbar for the chat mute list.

**Comment by kphoenix137** [comment]
> If we want to go beyond 8 players, though, we should probably take some of these suggestions seriously. Like turning the arrays into vectors and moving Golems to their own collection.

Wouldn't moving golems break reverse compatibility?

**Comment by AJenbo** [comment]
not if you map it

**Comment by AJenbo** [comment]
Proper support will require way more then is happening here, closing for now
